### ISSUE TITLE
No API to get module under code position

### ISSUE THREAD

**Issue by lukaszsamson** [elixir-lang/elixir#12767]
### Elixir and Erlang/OTP versions

any

### Operating system

any

### Current behavior

When implementing a debugger adapter it is necessary to be able to accurately discover what module a certain code position corresponds to. Module and line are required by erlang debugger when setting a breakpoint. In erlang this is easy becouse there's 1-1 mapping between .erl files and modules. In elixir one file can define many modules or can contain code that is not a part of any module.

The only API that somehow makes this possible is `token_metadata` from compiler options. With AST traversal one is able to get `do` and `end` positions and match against `:defmodule` etc calls.
The problem with that is it will only work for standard elixir code with standard AST. It will fail for DSLs. It will fail for cases when modules are macro generated.

The other candidate APIs do not provide necessary functionality:
- compile tracer `:on_module` only has module begin line in macro env
- EEP48 docs only register begin line and column
- runtime introspection via `__info__` and `module_info` only returns file to module mapping with no code ranges

Related issue https://github.com/elixir-lsp/elixir_sense/issues/94

### Expected behavior

Ability to match code position with a module

**Comment by josevalim** [comment]
>  The problem with that is it will only work for standard elixir code with standard AST. It will fail for DSLs. It will fail for cases when modules are macro generated.

Can you please provide more context here? Macro-generated code will still have do/end, no? Why would it also be different for a DSL?

**Comment by lukaszsamson** [comment]
> Macro-generated code will still have do/end, no? Why would it also be different for a DSL?

With macro code it's difficult to tell if something is a module. It's relatively easy to match on `{:defmodule, meta, args}` but it's not easy tell from AST that a macro will generate a module on compile. It's also difficult to correctly guess the number of defined modules from AST alone e.g.
```
for i <- 1..9 do
  defmodule :"Elixir.MyModule#{1}" do
  end
end
```

For elixirLS debugger I ended up with parsing the AST for `defmodule` and `defimpl` and a fallback to
```
:code.all_loaded()
        |> Enum.map(&elem(&1, 0))
        |> Enum.filter(fn module -> String.starts_with?(Atom.to_string(module), "Elixir.") end)
        |> Enum.group_by(fn module ->
          module_info = ModuleInfoCache.get(module) || module.module_info()
          Path.expand(to_string(module_info[:compile][:source]))
        end)
```
which returns all modules defined in a file. This way covers most cases but will fail in `for` example.
The alternative would be to attempt setting breakpoints in every module from file and not looking at AST at all but that feels wrong.

**Comment by josevalim** [comment]
Yes, this will also be solved as part of #12645, as we will expand code and be able to answer these questions :)
